src/osp/Active/SysNewton.cpp

```diff
@@ -352,32 +355,84 @@ ACompNwtWorld* SysNewton::try_get_physics_world(ActiveScene &rScene)
 }
 
 std::pair<ActiveEnt, ACompNwtBody*> SysNewton::find_rigidbody_ancestor(
-        ActiveScene& rScene, ActiveEnt ent)
+        ActiveScene& rScene, ActiveEnt ent, Matrix4* pTransform)
 {
     ActiveEnt prevEnt, currEnt = ent;
-    ACompHierarchy *currHeir = nullptr;
+    ACompHierarchy *currHier = nullptr;
 
     do
     {
+       currHier = rScene.get_registry().try_get<ACompHierarchy>(currEnt);
 
-       currHeir = rScene.get_registry().try_get<ACompHierarchy>(currEnt);
-
-        if (!currHeir)
+        if (!currHier)
         {
             return {entt::null, nullptr};
         }
+        
+        // If transform argument was passed, record the local transformation of
+        // the current node relative to its parent
+        if (pTransform && (currHier->m_level > gc_heir_physics_level))
+        {
+            Matrix4 localTransform = rScene.reg_get<ACompTransform>(currEnt).m_transform;
+            *pTransform = localTransform * (*pTransform);
+        }
 
         prevEnt = currEnt;
-        currEnt = currHeir->m_parent;
+        currEnt = currHier->m_parent;
     }
-    while (currHeir->m_level != gc_heir_physics_level);
+    while (currHier->m_level != gc_heir_physics_level);
 
-    auto *body = rScene.get_registry().try_get<ACompNwtBody>(prevEnt);
+    auto *body = rScene.get_registry().try_get<ACompRigidBody_t>(prevEnt);
 
     return {prevEnt, body};
 
 }
 
+std::pair<ACompNwtBody*, ACompTransform*> SysNewton::try_get_or_find_rigidbody_parent(
```

Comment:
I know this function name is kind of ridiculous but when you consider its function, it sort of fits. I'm open to any more concise ideas.

Reply:
I think this name is okay. It kind of makes sense, but passing an entity by reference seems a bit sketchy.

As a side note, multiple machine systems calling this function on their own for the same entity sounds more ridiculous. Perhaps a new component and system function can be added. An ACompRigidBodyAncestor managed by SysNewton that refers to an entity that has a rigid body, which can also track relative transform and stuff.

Reply:
Check the new version and see if you think it's an improvement